integration-tests/artemis-core/src/main/java/io/quarkus/it/artemis/ArtemisProducerManager.java

```diff
@@ -0,0 +1,35 @@
+package io.quarkus.it.artemis;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import org.apache.activemq.artemis.api.core.ActiveMQException;
+import org.apache.activemq.artemis.api.core.client.ClientMessage;
+import org.apache.activemq.artemis.api.core.client.ClientSession;
+import org.apache.activemq.artemis.api.core.client.ClientSessionFactory;
+import org.apache.activemq.artemis.api.core.client.ServerLocator;
+
+@ApplicationScoped
+public class ArtemisProducerManager {
+
+    @Inject
+    ServerLocator serverLocator;
+
+    private ClientSessionFactory connection;
+
+    @PostConstruct
+    public void init() throws Exception {
+        connection = serverLocator.createSessionFactory();
```

Comment:
I'm not quite sure whyw e don't enlist a CDI bean being the session factory provider in the extension

Reply:
Because a ClientSessionFactory should be regarded as connection. See [Artemis documentation](https://activemq.apache.org/components/artemis/documentation/latest/core.html#clientsessionfactory). If your connection breaks, the ClientSessionFactory should be created again.

Reply:
@middagj OK but that feels like a terrible developer experience.
The extension should make an effort to retry this connection itself. You can imagine very few users would think of doing it otherwise. Any advice on how to do it with CDI @mkouba

Reply:
Well, if I understand it correctly we could create bean that implements `ClientSessionFactory` and delegates to a factory created by `serverLocator.createSessionFactory()`. Then in every call we could verify the underlying factory and if something is wrong, recreate the factory again? @middagj @emmanuelbernard But I don't know anything about Artemis. So this is just a guess.

Reply:
I think the developer joy can be greatly improved. I would like to not have to worry about a connection at all and just want to define a message listener and have a producer bean which can be used to send new messages in the same transaction if configured. I don't know when all that would realistic to add on, and having the ability to use Artemis client in Quarkus already helps us and probably others already.

Sounds like a good solution. There are different session factories which can be created, but that can be put in a config.

Reply:
I was trying this approach, but I reconsider if this is the right way. `ClientSessionFactory ` implements `AutoCloseable` and it resembles the normal JMS Connection class. If you look at both the [JMS](2930/files#diff-3e8b5119c05a5414d227f6917f4071d3) and [Core](2930/files#diff-c0cc000a32a46da7f908389e2861bbf9) ArtemisProducerManager, they look quite similar. Both protocols have the distinction between a Connection and a Session, only Core protocol names the Connection a SessionFactory.

I think it would be better to focus on JMS and create something similar like Spring's [JmsListener](https://docs.spring.io/spring/docs/5.0.x/spring-framework-reference/integration.html#jms-receiving) and [JmsTemplate](https://docs.spring.io/spring/docs/5.0.x/spring-framework-reference/integration.html#jms-sending) to improve developer joy. I prefer to do it in a separate pull request as this request already enables people to add the dependency in a native image.